File: creational/dart/singleton_structured/async_singleton.dart
class AsyncSingleton {
  static AsyncSingleton? _instance;
  static bool _isInitializing = false;
  
  AsyncSingleton._internal() {
    print('AsyncSingleton: Instance created');
  }

  static Future<AsyncSingleton> getInstance() async {
    if (_instance != null) return _instance!;

    while (_isInitializing) {
      await Future.delayed(Duration(milliseconds: 10));
      if (_instance != null) return _instance!;
    }

    _isInitializing = true;
    try {
      print('AsyncSingleton: Starting heavy async initialization...');
      await Future.delayed(Duration(milliseconds: 100));
      _instance = AsyncSingleton._internal();
    } finally {
      _isInitializing = false;
    }

    return _instance!;
  }
}

File: creational/typescript/singleton_structured/async_singleton.ts
export class AsyncSingleton {
  private static instance: AsyncSingleton | null = null;
  private static isInitializing: boolean = false;
  
  private constructor() {
    console.log('AsyncSingleton: Instance created');
  }

  public static async getInstance(): Promise<AsyncSingleton> {
    if (AsyncSingleton.instance) return AsyncSingleton.instance;

    while (AsyncSingleton.isInitializing) {
      await new Promise(resolve => setTimeout(resolve, 10));
      if (AsyncSingleton.instance) return AsyncSingleton.instance;
    }

    AsyncSingleton.isInitializing = true;
    try {
      console.log('AsyncSingleton: Starting heavy async initialization...');
      await new Promise(resolve => setTimeout(resolve, 100));
      AsyncSingleton.instance = new AsyncSingleton();
    } finally {
      AsyncSingleton.isInitializing = false;
    }

    return AsyncSingleton.instance;
  }
}
